Update the "Optimization with SPSA" demo

dwierichs opened the pull request:
**Title:**
Optimization using SPSA

**Summary:**
This updates the existing demo with the same title.
- PennyLane's own `qml.SPSAOptimizer` is used instead of `noisyopt` optimization. This also reduces required reshaping hustle.
- Some other minor simplifications have been done. 
- A bug reporting smaller execution numbers for SPSA was fixed
- The text was updated corresponding to the newly used PL feature instead of `noisyopt`.

**Relevant references:**

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

josh146 commented on demonstrations/spsa.py:
```diff
@@ -171,86 +174,66 @@
-# required for completion using the callback function. This will be an
-# interesting quantity!
-from noisyopt import minimizeSPSA
+# We'll use the ``qml.SPSAOptimizer`` built into PennyLane, 200 iterations in total,
```
```suggestion
# We'll use the :class:`~pennylane.SPSAOptimizer` built into PennyLane, 200 iterations in total,
```

Hopefully this works 🤞

dwierichs replied:
Yes, this works @josh146  Thanks :)

dwierichs commented on demonstrations/spsa.py:
```diff
@@ -13,14 +13,15 @@
    tutorial_vqe_qng Accelerating VQEs with quantum natural gradient
 
-*Author: Antal Szava — Posted: 19 March 2021. Last updated: 13 September 2022.*
+*Author: Antal Szava — Posted: 19 March 2021. Last updated: xx February 2023.*
```
Should I modify the author info and add myself? Or is this still a minor edit?

josh146 replied:
Yep agreed!

timmysilv commented on demonstrations/spsa.py:
```diff
@@ -171,86 +169,80 @@
+        param = opt.step(fun, param)
+
+        # Print out the status of the optimization
+        if step % int(np.ceil(num_steps/20)) == 0:
```
the `int(np.ceil(num_steps/20))` loop invariant can be evaluated before the for loop

timmysilv commented on demonstrations/spsa.py:
```diff
@@ -284,63 +268,42 @@ def callback_fn(xk):
 #
 #  .. code-block:: none
 #
-#     Iteration = 10, Number of device executions = 14, Cost = 0.09
```
curious what changed here. used to be 1.5 device executions per iteration (with some minus-one shift), now it's 2. i'm mostly wondering about this because we are taking special care to point out the total device executions but it went up using the SPSAOptimizer!

josh146 replied:
Huh. Is noisy opt more efficient than ours?

dwierichs replied:
So the previous version of the demo divided the iteration count for spsa (and only for spsa!) by 2.
It had a [comment associated with that line](https://github.com/PennyLaneAI/qml/blob/e50d7b0483252d31f20296ae35fc2402caa77436/demonstrations/spsa.py#L232) stating that this was in order to correct for the execution of the device to monitor the cost function itself. However, that was incorrect, the correct mitigation for these evaluations is to subtract one device execution per iteration. 
SPSA uses two evaluations for each gradient estimate, so having `1.5*num_epochs` evaluations overall is not possible.
This is also seen further down where the text of the demo made the prognosis that we will need 6000 executions and then the number of executions reported by the log is smaller than that.

dwierichs replied:
Note that this bugfix gets us into trouble regarding the performance in the VQE example. I guess mostly because it has fewer parameters.

timmysilv replied:
but that's because we were.. lying before? so the new code isn't worse, just more honest?

dwierichs replied:
If I'm not entirely misunderstanding what was going on before, yes.

timmysilv commented on demonstrations/spsa.py:
```diff
@@ -414,109 +374,88 @@ def cost_gd(params):
 #  .. code-block:: none
 #
-#     Device execution ratio: 4.161375661375661
+#     Device execution ratio: 10.083
```
on the other hand... nice!

dwierichs replied:
As there was an outdated threshold in here, the 10 was wrong.
We're back to about 4 :D

timmysilv commented on demonstrations/spsa.py:
```diff
@@ -414,109 +374,88 @@ def cost_gd(params):
-# to optimize a real chemical system, that of the hydrogen molecule :math:`H_2`.
+# Now that we've explored the theoretical underpinnings of SPSA and its use for a
+# toy problem optimization, let's use it
+# to optimize a real chemical system, namely that of the hydrogen molecule :math:`H_2`.
```
I've heard the qchem folks call the H2 molecule a toy problem lol 😄 luckily I think their purposes differ sufficiently from an SPSA demo

timmysilv commented on demonstrations/spsa.py:
```diff
@@ -414,109 +374,88 @@ def cost_gd(params):
-    "qiskit.aer", wires=dev.num_wires, shots=1000, noise_model=noise_model
-)
+from qiskit.providers.aer import noise
+from qiskit.providers.fake_provider import FakeLima
```
was FakeLima just not available before? are we making this switch because it's easier on users?

rmoyard replied:
I think it is the new recommended way of getting the noise model

dwierichs replied:
The previous code did not execute on my machine, asking for IBM Q account things. I thought it's nicest if one does not have to bother with that just to get the noise model. Web search suggested that this indeed is the current way of obtaining the noise model :)

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -171,86 +169,80 @@
+# number of device executions along the way. The latter will be an interesting
+# quantity to evaluate the optimization cost on hardware!
+
+def run_optimizer(opt, fun, init_param, num_steps):
```
Minor: Just to make it a be more clear; you could name the arg `optimizer` and `cost_function`

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -284,63 +268,42 @@ def callback_fn(xk):
 
-def cost_gd(params):
-    return qnode_gd(params)
+# Create a new device and qnode
```
```suggestion
# Create a new device and a qnode as cost function
```

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -398,14 +357,15 @@ def cost_gd(params):
-grad_desc_exec_min = device_execs_grad[np.argmin(cost_store_grad)]
-spsa_exec_min = device_execs_spsa[np.argmin(cost_store_spsa)]
-print(f"Device execution ratio: {grad_desc_exec_min/spsa_exec_min}.")
+grad_execs_to_prec = exec_history_grad[np.where(np.array(cost_history_grad) < 0.800)[0][0]]
```
Where does the 0.800 come from?

dwierichs replied:
Oh, this is outdated. Thanks!

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -525,23 +464,32 @@ def exp_val_circuit(params):
-#
-#     Final estimated value of the ground-state energy = -1.07234725 Ha
-#     Accuracy with respect to the true energy: 0.06384221 Ha
+# Running the GradientDescentOptimizer optimizer for 15 iterations.
```
Given the numbers it might be not worth to display all of them

dwierichs replied:
Went back to every 5th.

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -635,47 +543,40 @@ def callback_fn(xk):
-#
-#     Final estimated value of the ground-state energy = -1.06069987 Ha
-#     Accuracy with respect to the true energy: 0.07548958 Ha
+# Running the SPSAOptimizer optimizer for 200 iterations.
```
Here you also reduce the number of executions, as you have more executions per iteration

dwierichs replied:
This is actually wrong again, I have patched the `run_optimizer` function to account for Hamiltonian decompositions.

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -739,6 +638,12 @@ def callback_fn(xk):
+#    Stochastic Approximation of the Quantum Fisher Information",
+#    `Quantum, 5, 567 <https://quantum-journal.org/papers/q-2021-10-20-567/>`__, Oct 2021
 
 ##############################################################################
```
https://github.com/PennyLaneAI/qml/blob/71dc81cc508036b7d8474adf3825ed37fa321f21/demonstrations/spsa.py#L588

I would suggest to match the number of executions on this graph

dwierichs replied:
We're now at 4800 (SPSA) vs 4995 (GradDesc).

rmoyard commented on demonstrations/spsa.py:
```diff
@@ -739,6 +638,12 @@ def callback_fn(xk):
+#    `Quantum, 5, 567 <https://quantum-journal.org/papers/q-2021-10-20-567/>`__, Oct 2021
 
 ##############################################################################
 # About the author
```
You can add yourself here :100:

rmoyard reviewed (commented):
Great job @dwierichs :100: I have left some comments and suggestions

dwierichs commented on demonstrations/spsa.py:
```diff
@@ -13,14 +13,16 @@
-*Author: Antal Szava — Posted: 19 March 2021. Last updated: 13 September 2022.*
+*Author: Antal Szava — Posted: 19 March 2021. Last updated: xx February 2023.*
 
-In this tutorial, we investigate using a gradient-free optimizer called
```
Ah, no, we're still using the `SPSAOptimizer` built into PennyLane, replacing `noisyopt`.
However, I find it a bit confusing to say "gradient-free" because `SPSA` _does_ estimate the gradient. 
I mean, it is not an analytic gradient but a stochastic estimate, but that's true in quantum in many other settings.
If `SPSA` is "gradient-free", by that logic, gradient descent with a shot-based simulator also is "gradient-free" :grimacing:

dwierichs commented on demonstrations/spsa.py:
```diff
@@ -171,86 +169,80 @@
-)
+ansatz = qml.StronglyEntanglingLayers
 
+all_pauliz_tensor_prod = qml.operation.Tensor(*[qml.PauliZ(i) for i in range(num_wires)])
```
Ah, I'll check :+1:

dwierichs replied:
Works.

dwierichs commented:
Given that we switch the IBMQ import to a `fake_providers` noise model, one could just have this tutorial run like the others, without hardcoding outputs and plots.
On my laptop it took about 8 minutes to run in its latest form.
@josh146  @rmoyard @timmysilv what do you think?

josh146 commented:
> Given that we switch the IBMQ import to a fake_providers noise model, one could just have this tutorial run like the others, without hardcoding outputs and plots.

I am in favour!

rmoyard commented:
> Given that we switch the IBMQ import to a fake_providers noise model, one could just have this tutorial run like the others, without hardcoding outputs and plots.

@dwierichs I am also in favor, can we move it to tutorial then? It will be run automatically and remove some burden on the core team :)

dwierichs commented:
I moved it to `tutorial_spsa.py`. The one main issue that needs addressing is that SPSA does not outperform Gradient descent anymore, now that we updated the way we count the executions.
One could test with an enhanced ansatz, trying to get better energies for both optimizers.

dwierichs commented:
>The one main issue that needs addressing is that SPSA does not outperform Gradient descent anymore, now that we updated the way we count the executions.

"fixed" this by using a different ansatz.

rmoyard reviewed (approved):
Thanks it looks good to me :100: Very great update and good catch for the ansatz!

timmysilv reviewed (approved):
love it, thanks for getting this done! don't forget to update the date on the demo. also, should we add QNSPSA to the Related section for SPSA? I'm fine either way, just figured since QNSPSA relates to SPSA.

dwierichs commented:
I was wondering about this as well. I thought it's not a causal dependence or anything like that, so having a cycle of ref links seems ok :)

dwierichs commented on demonstrations/tutorial_spsa.py:
```diff
@@ -0,0 +1,555 @@
+   tutorial_vqe_qng Accelerating VQEs with quantum natural gradient
+   qnspsa Quantum natural SPSA optimizer
+
+*Authors: Antal Szava & David Wierichs — Posted: 19 March 2021. Last updated: xx February 2023.*
```
```suggestion
*Authors: Antal Szava & David Wierichs — Posted: 19 March 2021. Last updated: 10 February 2023.*
```
